### PR TITLE
Boilerplate Changes for MWS Project

### PR DESCRIPTION
**Pull request by bluemellophone**
This PR adds the ability to run the existing Codex application alongside the MWS application.

To test with the MWS activation, run the following:

```bash
git fetch
git checkout mws-intro
git pull
./scripts/mws/activate.sh
pytest
```

To reset with the Codex activation

```bash
./scripts/codex/activate.sh
```

## Pull Request Overview

- Added application namespace for the existing functionality, configurations, and tasks for `codex`.  This change moves the following files:
  - `.env` to `.env.codex`
  - `_frontend/` to `_frontend.codex/` [submodule]
  - `config.py` to `config.codex.py`
  - `Dockerfile` to `Dockerfile.codex`
  - `docker-compose.yml` to `docker-compose.codex.yml`
  - `invoke app.*` to `invoke codex.*` [invoke tasks]
  - Renamed the Codex Docker image from `wildme/houston:latest` to `wildme/codex:latest`
  - Renamed the Python virtual environment from `houston3.7` to `codex`
  - `.dockerfiles/docker-entrypoint-init.d/_set_up_gitlab.py` to `.dockerfiles/docker-entrypoint-init.d.codex/_set_up_gitlab.py`
- Added an additional application namespace for `mws` for the Modernizing Wildlife Surveys (MWS) project.  The change adds the following files:
  - `.env.mws`
  - `_frontend.mws/` [submodule]
  - `config.mws.py`
  - `Dockerfile.mws`
  - `docker-compose.mws.yml`
  - `invoke mws.*`
- Added symlinks for proper runtime usage
  - `.env`
  - `config.py`
  - `docker-compose.yml`
  - Running `./scripts/codex/activate.sh` will set the development environment to the Codex activation
  - Running `./scripts/mws/activate.sh` will set the development environment to the MWS activation
- Added `missions` module and database update `922f78ec3698`, added `tests/modules/missions`
- Added namespacing for all scripts:
  - `scripts/codex/`
  - `scripts/mws/`
  - `scripts/docs/`
  - `scripts/swagger/` (`./scripts/build.swagger-ui.sh` moved to `./scripts/swagger/build.frontend.sh`)
  - `scripts/tests/` (`./scripts/run_tasks_for_coverage.sh` moved to `./scripts/tests/run_tasks_for_coverage.sh`)
  - `scripts/utils/` (`./scripts/publish.sh` moved to `./scripts/utils/publish.sh`)
  - All `scripts/host.*` scripts and `scripts/test.sh` have been deleted
- Partitioned extensions into required and optional, optional extensions must be enabled explicitly in `setup.py`.  The new optional extensions are:
  - `acm`
  - `edm`
  - `cors`
  - `elasticsearch`
  - `gitlab`
  - `tus`
  - `mail`
  - `stripe`
- Added new wrappers to turn off functionality and explicitly label required modules for shared functionality (see `flask_restx_patched/namespace.py`)
  - Added Python function/method decorators `@module_required` and `@extension_required`
  - Added Flask REST API decorators `@api.module_required` and `@api.extension_required`
  - Added functions `app.modules.is_module_enabled` and `app.extensions.is_extension_enabled`
  - Added `pytest.mark.skipif` to non-MWS module tests
- Added switched behavior for Elastic Search and the tasks and indices it builds on launch
- ~~Fixed GitHub action for CI test `Github Cleanup`~~
- Upgraded `db.relationships(back_populates)` between `User` and non-MWS modules and models to `db.relationships(backref)` where needed
- Added a test (thanks @karenc) to ensure the `order_by` ordering of `backref` relationships is maintained
- Added testing for MWS activation
- Added `confirm_deleted_rows: False` to fix test warnings with models `Annotation`, `AssetGroupSighting`, `AssetGroup`, `CollaborationUserAssociations`, `Collaboration`, `Encounter`, `FileUpload`, `SightingAssets`, `SiteSetting`
- Fixed test warnings that were holding on to file descriptors and leaking open IO handlers
- Renamed `TestCreationData` to `AssetGroupCreationData` to prevent `pytest` from reading `Test*` class names as test conditions
- Converted `logging.warn` to `logging.warning` to fix test deprecation warnings
- Updated `setup.cfg` to enable `skip-string-normalization` in Brunette, allowing the `double-quote-string-fixer` `pre-commit` hook to standardize string normalizations.

---

**Review Notes**

The decorators `@module_required` and `@extension_required` are available for object methods and for REST API resource endpoints.  See examples below:

```python
# EXAMPLE 1

# If projects is disabled, `user_obj.get_projects` will log a warning message and return `[]`
@module_required('projects', resolve='warn', default=[])
def get_projects(self):
    return [enrollment.project for enrollment in self.project_membership_enrollments]

# EXAMPLE 2

def default_func(obj):
    return [None]

# You may use a function as the default (inputs to the wrapped function are passed to the default function)
@module_required('projects', resolve='warn', default=default_func)
def get_projects(self):
    return [enrollment.project for enrollment in self.project_membership_enrollments]

# EXAMPLE 3

# Ypu may also pass the required mpdules as an iterable.  
# The default resolve is `error` and will throw an explicit RuntimeError if the function is called and the required modules are disabled
@module_required(('encounters', 'annotations'))
def get_my_annotations(self):
    annotations = []
    for encounter in self.owned_encounters:
        annotations.extend(encounter.annotations)
    return annotations

# EXAMPLE 4

# If the EDM extension is disabled, the user will receive a `HTTPStatus.NOT_IMPLEMENTED` error response
@api.route('/edm/sync')
@api.extension_required('edm')
class UserEDMSync(Resource):
    ...

# EXAMPLE 5

# Tests can also be disabled by requiring modules to be enabled for the test to proceed
from tests.utils import module_unavailable
@pytest.mark.skipif(
    module_unavailable('individuals'), reason='Individuals module disabled'
)
def test_get_set_individual_names(db, flask_app_client, researcher_1):
    ...
```

**Model Updates**

Missions
- `guid`
- `title`
- `user_membership_enrollments`
- `owner_guid`
- `owner`

MissionUserMembershipEnrollment
- `mission_guid`
- `user_guid`
- `mission`
- `user`

**Database Updates**

Database migration `922f78ec3698` to add the `mission` and `mission_user_membership_enrollment` tables.

### CODE REVIEW

File: scripts/test.sh [hunk 1]
```diff
@@ -1,6 +0,0 @@
-#!/bin/bash
```
Comment (bluemellophone) [inline scripts/test.sh:1]: Note that this file is not needed for automated testing

File: app/modules/assets/schemas.py [hunk 2]
```diff
@@ -30,14 +30,18 @@ class DetailedAssetSchema(BaseAssetSchema):
     Detailed Asset schema exposes all useful fields.
     """
 
+    # <TODO - MWS>
```
Comment (bluemellophone) [inline app/modules/assets/schemas.py:33]: There is a needed change (marked as `<TODO -MWS>`) to the asset schema to disentangle it with asset_groups.  The solution I have for now is to patch out the value from the schema.  Needs a permanent solution

File: app/modules/encounters/models.py [hunk 3]
```diff
@@ -36,15 +36,29 @@ class Encounter(db.Model, FeatherModel):
...
+    # owner = db.relationship(
+    #     'User', back_populates='owned_encounters', foreign_keys=[owner_guid]
+    # )
     owner = db.relationship(
-        'User', back_populates='owned_encounters', foreign_keys=[owner_guid]
+        'User',
+        backref=db.backref('owned_encounters'),
+        foreign_keys=[owner_guid],
+        primaryjoin='User.guid == Encounter.owner_guid',
+        order_by='Encounter.guid',
```
[the remark this answers is not in the document]
Comment (colinwkingen) [reply app/modules/encounters/models.py:47]: Awesome, thanks for tackling this Karen.

File: setup.cfg [hunk 4]
```diff
@@ -44,7 +44,7 @@ exclude = .git,.env,migrations
 [tool:brunette]
 line-length = 90
 verbose = false
-# skip-string-normalization = true
+skip-string-normalization = true
 single-quotes = true
```
Comment (colinwkingen) [inline setup.cfg:48]: Huh, wonder why this was off.

Comment (colinwkingen) [reply setup.cfg:48]: Wait, you are turning it off. Why is that?

File: setup.py [hunk 5]
```diff
@@ -55,7 +55,7 @@ def _minimal_ext_cmd(cmd):
 MAINTAINER_EMAIL = 'dev@wildme.org'
 DESCRIPTION = 'The backend server for the new Wildbook frontend and API.'
 LONG_DESCRIPTION = DESCRIPTION
-KEYWORDS = ['wild me', 'hoston', 'wildbook']
+KEYWORDS = ['wild me', 'houston', 'wildbook', 'codex', 'mws']
```
Comment (colinwkingen) [inline setup.py:58]: Heh, cool.

**Review by colinwkingen — COMMENTED** [review]
Looks good. Thanks for all the attention put to correcting comments and such during this process in addition to the code and script changes.

File: tests/modules/missions/resources/utils.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (bluemellophone) [reply tests/modules/missions/resources/utils.py:1]: I saw the updated interface.  I'm going to use the boilerplate copied from `projects` for now and we can upgrade all of them in a separate PR down the line

**Pull request merged**